[dynamo 3.11] implement 3.11 exceptiontable

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99180
* __->__ #96511

Summary of changes:
- Add CPython exceptiontable parsing/assembling functions in torch/_dynamo/bytecode_transformation.py, based on https://github.com/python/cpython/blob/3.11/Objects/exception_handling_notes.txt.
- Add optional `exn_tab_entry` field to dynamo `Instruction`s in torch/_dynamo/bytecode_transformation.py in order to virtualize exception table entries (start, end, target instructions).
- Add checks guarding against duplicate instructions in dynamo, so that jump/exceptiontable targets are unambiguous. See `get_indexof` in torch/_dynamo/bytecode_analysis.py. Ensure that bytecode generation throughout dynamo does not generate duplicate instructions.
- Allow dynamo bytecode generation logic to generate nested exception table entries for developer convenience. CPython expects entries to not overlap, so we flatten nested entries during assembly in torch/_dynamo/bytecode_transformation.py:compute_exception_table.
- Simulate the block stack in torch/_dynamo/symbolic_convert.py. CPython removed the block stack in 3.11, but dynamo needs it in order to keep track of active contexts. So we simulate the block stack as before by looking at exceptiontable entries in order to determine the current blocks.
- Update context codegen in torch/_dynamo/resume_execution.py. The `SETUP_FINALLY` bytecode, which conveniently had a jump target to the finally block, was removed in 3.11, so we need to keep track of the jump target of the finally block using exceptiontables. Generating resume functions is more difficult since the original exceptiontable entries pointing to old cleanup code need to be modified to point to new cleanup code.
- Fix a push_null bug in torch/_dynamo/variables/functions.py introduced by https://github.com/pytorch/pytorch/pull/98699

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @soumith @anijain2305 @Xia-Weiwen @peterbell10

williamwen42 committed:
[dynamo 3.11] implement 3.11 exceptiontable

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[WIP, dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on torch/_dynamo/bytecode_transformation.py:
```diff
@@ -289,6 +312,81 @@ def _update(delta, size):
+
+def encode_exn_tab_varint(n):
+    # unfortunately cannot use encode_varint since the order of
+    # the bytes is reversed for exception tables.
```
Add docstring explaining that this encoding is, use full words for function names.

jansel commented on torch/_dynamo/bytecode_transformation.py:
```diff
@@ -289,6 +312,81 @@ def _update(delta, size):
+    return b
+
+
+def decode_exn_tab_varint(bytes_iter):
```
docstring and use full words in function names

williamwen42 committed:
Update on "[dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] implement 3.11 exceptiontable"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

albanD commented on torch/_dynamo/bytecode_transformation.py:
```diff
@@ -402,6 +533,189 @@ def devirtualize_jumps(instructions):
+    """
+    Checks `tab` is a properly sorted list of nested InstructionExnTabEntry's,
+    i.e. no entries partially overlap.
+    "Properly sorted" means entries are sorted by increasing starts, then
```
I'm a bit confused by that. If there are no partial overlap why does it matter that they are sorted by decreasing ends?

williamwen42 replied:
It's so that when we iterate through sorted entries, any deeper nested entries that have the same start will appear later.

williamwen42 commented on torch/_dynamo/bytecode_analysis.py:
```diff
@@ -24,9 +25,23 @@
+    indexof = {}
+    for i, inst in enumerate(insts):
+        if id(inst) in indexof:
+            breakpoint()
```
Breakpoint should be removed. Hitting this line means we did something unintentional in dynamo.

williamwen42 commented on torch/_dynamo/bytecode_analysis.py:
```diff
@@ -24,9 +25,23 @@
+        if id(inst) in indexof:
+            breakpoint()
+        assert id(inst) not in indexof
+        indexof[id(inst)] = i
```
Most places that this code replaces use `id(inst)` instead. My intuition is that we want to make sure each instruction object gets its own key in case the equality/hash functions change in the future. @jansel do you recall the original intent?

williamwen42 commented on torch/_dynamo/resume_execution.py:
```diff
@@ -67,39 +79,79 @@ def try_except(self, code_options, cleanup: List[Instruction]):
+                *create_reset(),
                 create_instruction("JUMP_FORWARD", target=cleanup_complete_jump_target),
                 except_jump_target,
-                *reset,
```
It was ok in the past, but we now enforce instruction lists to not have duplicates so that jump/exceptiontable targets are unambiguous.

williamwen42 commented on torch/_dynamo/resume_execution.py:
```diff
@@ -187,16 +268,45 @@ def create_load_none():
+    """
+    it = iter(l2)
+    res = []
+    try:
```
This does not replace 2 nested for loops - we iterate through `l1` and we have a second pointer pointing to `l2` and only increment that pointer. So this algorithm is linear time.

williamwen42 commented on torch/_dynamo/output_graph.py:
```diff
@@ -539,7 +539,7 @@ def compile_subgraph(
                         )
                     )
                 else:
-                    prefix_insts.append(inst)
```
Instructions in dynamo instruction lists need to be unique, so a copy is required. See the graph break logic in InstructionTranslatorBase.step to see an example of when the instructions may not be unique without a copy.

williamwen42 commented on torch/_dynamo/bytecode_transformation.py:
```diff
@@ -289,6 +322,93 @@ def _update(delta, size):
+    b = []
+    for entry in tab:
+        first_entry = encode_exception_table_varint(entry.start // 2)
+        first_entry[0] |= 1 << 7
```
According to https://github.com/python/cpython/blob/3.11/Objects/exception_handling_notes.txt:
> We need a marker for the start of the entry, so the first byte of entry will have the most significant bit set.

williamwen42 committed:
Update on "[dynamo 3.11] implement 3.11 exceptiontable"


As part of the exceptiontable implementation, we now enforce that dynamo generates instruction lists without duplicates before assembly.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo 3.11] implement 3.11 exceptiontable"


Summary of changes:
- Add CPython exceptiontable parsing/assembling functions in torch/_dynamo/bytecode_transformation.py, based on https://github.com/python/cpython/blob/3.11/Objects/exception_handling_notes.txt.
- Add optional `exn_tab_entry` field to dynamo `Instruction`s in torch/_dynamo/bytecode_transformation.py in order to virtualize exception table entries (start, end, target instructions).
- Add checks guarding against duplicate instructions in dynamo, so that jump/exceptiontable targets are unambiguous. See `get_indexof` in torch/_dynamo/bytecode_analysis.py. Ensure that bytecode generation throughout dynamo does not generate duplicate instructions.
- Allow dynamo bytecode generation logic to generate nested exception table entries for developer convenience. CPython expects entries to not overlap, so we flatten nested entries during assembly in torch/_dynamo/bytecode_transformation.py:compute_exception_table.
- Simulate the block stack in torch/_dynamo/symbolic_convert.py. CPython removed the block stack in 3.11, but dynamo needs it in order to keep track of active contexts. So we simulate the block stack as before by looking at exceptiontable entries in order to determine the current blocks.
- Update context codegen in torch/_dynamo/resume_execution.py. The `SETUP_FINALLY` bytecode, which conveniently had a jump target to the finally block, was removed in 3.11, so we need to keep track of the jump target of the finally block using exceptiontables. Generating resume functions is more difficult since the original exceptiontable entries pointing to old cleanup code need to be modified to point to new cleanup code.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

williamwen42 commented:
@pytorchbot merge -ic
